### PR TITLE
TF2 doc: api.py docstring review

### PR DESCRIPTION
**Pull request by jimthompson5802**
# Documentation Pull Requests

Reviewed and updated docstring for `ludwig.api.k_fold_cross_validate()` function.

Before I do the remaining docstring reviews in this module, let me know if the form of the `k_fold_cross_validate()` docstrings looks correct.

### CODE REVIEW

**Comment by jimthompson5802** [comment]
I refactored the k_fold parameter spec to match current design. i.e., changed `data_csv` to `dataset` and `data_format` parameters.  To support the new data formats, I created in `ludwig.utils.data_utils.external_data_reader_registry` a new registry explicit for the readers of the new data formats.  One limitation, `hdf5` is not support for k_fold processing right now because it's reader does not return a pandas data frame.  if there is a desire to support `hdf5`, I'll have to do some research to figure out how to make the modifications.

Updated the unit test to reflect new parameter specification and added test for using the new data formats.  Updated the k_fold example code to reflect new parameter specification.

Local run of `test_kfold_cv.py`
```
=========================================================== test session starts ============================================================
platform linux -- Python 3.6.9, pytest-6.0.2, py-1.9.0, pluggy-0.13.1 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /opt/project
plugins: forked-1.3.0, pycharm-0.7.0, xdist-2.1.0, typeguard-2.9.1
collected 21 items

../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_cli[features_to_use0] PASSED                                           [  4%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_cli[features_to_use1] PASSED                                           [  9%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_cli[features_to_use2] PASSED                                           [ 14%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_cli[features_to_use3] PASSED                                           [ 19%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_cli[features_to_use4] PASSED                                           [ 23%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_cli[features_to_use5] PASSED                                           [ 28%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_api_from_file PASSED                                                   [ 33%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_api_in_memory PASSED                                                   [ 38%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[csv] PASSED                                            [ 42%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[df] PASSED                                             [ 47%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[dict] PASSED                                           [ 52%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[excel] PASSED                                          [ 57%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[feather] PASSED                                        [ 61%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[fwf] PASSED                                            [ 66%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[html] PASSED                                           [ 71%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[json] PASSED                                           [ 76%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[jsonl] PASSED                                          [ 80%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[parquet] PASSED                                        [ 85%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[pickle] PASSED                                         [ 90%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[stata] PASSED                                          [ 95%]
../../tests/integration_tests/test_kfold_cv.py::test_kfold_cv_dataset_formats[tsv] PASSED                                            [100%]

```

File: ludwig/api.py [hunk 1]
```diff
@@ -138,75 +138,91 @@ def train(
...
+        :param training_set: (Union[str, dict, pandas.DataFrame], default: `None`)
+            source containing training data.
+        :param validation_set: (Union[str, dict, pandas.DataFrame], default: `None`)
+            source containing validation data.
+        :param test_set: (Union[str, dict, pandas.DataFrame], default: `None`)
+            source containing test data.
+        :param training_set_metadata: (Union[str, dict], default: `None`)
+            metadata JSON file or loaded metadata.
+            Intermediate preprocess structure containing the mappings of the input
+            CSV created the first time a CSV file is used in the same
```
Comment (w4nderlust) [inline ludwig/api.py:163]: input CSV -> input dataset

Comment (w4nderlust) [reply ludwig/api.py:163]: Csv file -> dataset file
.json -> . meta.json

File: ludwig/api.py [hunk 2]
```diff
@@ -138,75 +138,91 @@ def train(
...
+            directory with the same name and a '.json' extension.
+        :param data_format: (str, default: `None`) format to interpret data sources. Will be inferred
                automatically if not specified.
```
Comment (w4nderlust) [inline ludwig/api.py:166]: should add a list of supported formats: csv, tsv, parquet...

File: ludwig/api.py [hunk 3]
```diff
@@ -138,75 +138,91 @@ def train(
...
+            directory with the same name and a '.json' extension.
+        :param data_format: (str, default: `None`) format to interpret data sources. Will be inferred
+               automatically if not specified.
+        :param experiment_name: (str, default: `'experiment'`) name for
+            the experiment.
+        :param model_name: (str, default: `'run'`) name of the model that is
+            being used.
+        :param model_resume_path: (str, default: `None`) resumes training of
+            the model from the path specified. The difference with
+            model_load_path is that also training statistics like the current
```
Comment (w4nderlust) [inline ludwig/api.py:191]: two times "also", the state of the optimizer is another thing that is resumed, should list it

File: ludwig/api.py [hunk 4]
```diff
@@ -138,75 +138,91 @@ def train(
...
-               the weights and just find out what performance can a model get
-               with a set of hyperparameters, use this parameter to skip it,
-               but the model will not be loadable later on.
+            saving model weights and hyperparameters each time the model
+            improves. By default Ludwig saves model weights after each epoch
+            the validation metric improves, but if the model is really big
+            that can be time consuming if you do not want to keep
+            the weights and just find out what performance can a model get
+            with a set of hyperparameters, use this parameter to skip it,
+            but the model will not be loadable later on.
```
Comment (w4nderlust) [inline ludwig/api.py:205]: Add at the end: and the returned model will have the weights obtained at the end of training, instead of the weights of the epoch with the best validation performance.

File: ludwig/api.py [hunk 5]
```diff
@@ -138,75 +138,91 @@ def train(
...
+            after each epoch for enabling resuming of training, but if
+            the model is really big that can be time consuming and will uses
+            twice as much space, use this parameter to skip it, but training
+            cannot be resumed later on.
+        :param skip_save_log: (bool, default: `False`) disables saving
+            TensorBoard logs. By default Ludwig saves logs for the TensorBoard,
+            but if it is not needed turning it off can slightly increase the
+            overall speed.
+        :param skip_save_processed_input: (bool, default: `False`) if a CSV
+            dataset is provided it is preprocessed and then saved as an hdf5
```
Comment (w4nderlust) [inline ludwig/api.py:217]: CSV dataset -> dataset file

File: ludwig/api.py [hunk 6]
```diff
@@ -138,75 +138,91 @@ def train(
...
+            and json to avoid running the preprocessing again. if this
+            parameter is False, the hdf5 and json file are not saved.
+        :param output_directory: (str, default: `'results'`) the directory that
+            will contain the training statistics, the saved model and
+            the training progress files.
```
Comment (w4nderlust) [inline ludwig/api.py:222]: add the TensorBoard logs

File: ludwig/api.py [hunk 7]
```diff
@@ -429,25 +448,35 @@ def train(
...
 
-    def train_online(self,
-                     dataset,
-                     training_set_metadata=None,
-                     data_format='auto',
-                     random_seed=default_random_seed,
-                     debug=False):
+    def train_online(
+            self,
+            dataset: str,
```
Comment (w4nderlust) [inline ludwig/api.py:455]: dataset is not only string

File: ludwig/api.py [hunk 8]
```diff
@@ -429,25 +448,35 @@ def train(
...
+            training_set_metadata: Union[str, dict] = None,
+            data_format: str = 'auto',
+            random_seed: int = default_random_seed,
+            debug: bool = False
+    ) -> None:
         """Performs one epoch of training of the model on `dataset`.
 
+        #Inputs
+
         :param dataset: (string, dict, DataFrame) source containing the training dataset.
```
Comment (w4nderlust) [inline ludwig/api.py]: copy the dataset entry from training

File: ludwig/api.py [hunk 9]
```diff
@@ -487,16 +516,51 @@ def train_online(self,
...
+            training progress files.
+        :param return_type: (Union[dict, pandas.DataFrame], default: pd.DataFrame)
+            indicates the format to
+            return predictions.
+        :param debug: (boolean, default: `False`) If `True` turns on `tfdbg`
+                with `inf_or_nan checks`.
+
+        # Return
+
+        :return: (Tuple[Union[dict, pd.DataFrame], str]) `(postproc_predictions, output_directory)`
```
Comment (w4nderlust) [inline ludwig/api.py:561]: postproc_predictions -> predictions

File: ludwig/api.py [hunk 10]
```diff
@@ -575,54 +639,58 @@ def evaluate(
...
+        :param return_type: (Union[dict, pandas.DataFrame], default: pandas.DataFrame) indicates
+            the format to
+            return predictions.
+        :param debug: (boolean, default: `False`) If `True` turns on `tfdbg`
+                with `inf_or_nan` checks.
 
 
         # Return
-
+        :return: (`stats`, `postprocess_predictions`, `output_directory`)
```
Comment (w4nderlust) [inline ludwig/api.py:692]: postprocessed_predictions - predictions

File: ludwig/api.py [hunk 11]
```diff
@@ -708,36 +776,137 @@ def evaluate(
...
+            to use multithreading parallelism to improve performance at
+            the cost of determinism.
+        :param use_horovod: (bool, default: `None`) flag for using horovod.
+        :param random_seed: (int: default: 42) random seed used for weights
+            initialization, splits and any other random function.
+        :param debug: (bool, default: `False) if `True` turns on `tfdbg` with
+            `inf_or_nan` checks.
+
+        # Return
+        :return: (Tuple[dict, dict, tuple, str)) `(test_results, train_stats, preprocessed_data, output_directory)`
```
Comment (w4nderlust) [inline ludwig/api.py:903]: test_results -> evaluation_statistics
train_stats -> training_statistics
overall statistics -> evaluation performance statistics
overall training statistics -> training performance statistics per epoch

File: ludwig/api.py [hunk 12]
```diff
@@ -216,10 +232,13 @@ def train(
 
         # Return
 
-        :return: ((dict, DataFrame)) tuple containing:
-            - A dictionary of training statistics for each output feature containing
-              loss and metrics values for each epoch. The second return
-            - A Pandas DataFrame of preprocessed training data.
+        :return: (Tuple[dict, Union[dict, pd.DataFrame], str]) tuple containing
+            `(train_stats, preprocessed_data, output_directory)`.
+            `train_stats` is a dictionary of training statistics for each
```
Comment (w4nderlust) [inline ludwig/api.py:239]: train_stats -> train_statistics

File: ludwig/api.py [hunk 13]
```diff
@@ -993,22 +1193,40 @@ def save(self, save_path):
         )
         save_json(training_set_metadata_path, self.training_set_metadata)
 
-    def save_model_definition(self, save_path):
+    def save_model_definition(
+        self,
+        save_path: str
+    ):
```
Comment (w4nderlust) [inline ludwig/api.py:1201]: add return type annotation -> None

**Comment by w4nderlust** [comment]
Added a few comment. Note that some of them apply every time a specific parameter is used, did not repeat them every time.

File: ludwig/api.py (hunk 2 above)
Comment (jimthompson5802) [reply ludwig/api.py:166]: done

File: ludwig/api.py (hunk 3 above)
Comment (jimthompson5802) [reply ludwig/api.py:191]: done

File: ludwig/api.py (hunk 4 above)
Comment (jimthompson5802) [reply ludwig/api.py:205]: done.

File: ludwig/api.py (hunk 1 above)
Comment (jimthompson5802) [reply ludwig/api.py:163]: done

File: ludwig/api.py (hunk 5 above)
Comment (jimthompson5802) [reply ludwig/api.py:217]: done

File: ludwig/api.py (hunk 6 above)
Comment (jimthompson5802) [reply ludwig/api.py:222]: done

File: ludwig/api.py (hunk 8 above)
Comment (jimthompson5802) [reply ludwig/api.py]: done

File: ludwig/api.py (hunk 9 above)
Comment (jimthompson5802) [reply ludwig/api.py:561]: done

File: ludwig/api.py (hunk 11 above)
Comment (jimthompson5802) [reply ludwig/api.py:903]: done.

File: ludwig/api.py (hunk 12 above)
Comment (jimthompson5802) [reply ludwig/api.py:239]: done

File: ludwig/api.py (hunk 13 above)
Comment (jimthompson5802) [reply ludwig/api.py:1201]: done

File: ludwig/api.py (hunk 7 above)
Comment (jimthompson5802) [reply ludwig/api.py:455]: done

File: ludwig/api.py [hunk 14]
```diff
@@ -162,8 +162,13 @@ def train(
             Intermediate preprocess structure containing the mappings of the input
             CSV created the first time a CSV file is used in the same
             directory with the same name and a '.json' extension.
-        :param data_format: (str, default: `None`) format to interpret data sources. Will be inferred
-               automatically if not specified.
+        :param data_format: (str, default: `None`) format to interpret data
```
Comment (w4nderlust) [inline ludwig/api.py:165]: I know I'm being pedantic... but I would put backticks before each format in the list, so that the commas are outside of the backticked part

File: ludwig/api.py [hunk 15]
```diff
@@ -197,10 +197,11 @@ def train(
         :param model_name: (str, default: `'run'`) name of the model that is
             being used.
         :param model_resume_path: (str, default: `None`) resumes training of
-            the model from the path specified. The difference with
-            model_load_path is that also training statistics like the current
-            epoch and the loss and performance so far are also resumed
-            effectively continuing a previously interrupted training process.
+            the model from the path specified. The model definition is restored.
+            In addition to model definition, training statistics and loss for
```
Comment (w4nderlust) [inline ludwig/api.py:201]: and los and the state, would replace with a comma before the loss

File: ludwig/api.py [hunk 16]
```diff
@@ -930,10 +934,12 @@ def experiment(
             `inf_or_nan` checks.
 
         # Return
-        :return: (Tuple[dict, dict, tuple, str)) `(test_results, train_stats, preprocessed_data, output_directory)`
-            `test_results` dictionary with
-            overall statistcs on the test_set, `train_stats` dictionary with
-            overall training statistics,
+        :return: (Tuple[dict, dict, tuple, str)) `(evaluation_results, training_stats, preprocessed_data, output_directory)`
+            `evaluation_results` dictionary with evaluation performance
```
Comment (w4nderlust) [inline ludwig/api.py:942]: evaluation_results -> evaluation_statistics

File: ludwig/api.py [hunk 17]
```diff
@@ -930,10 +934,12 @@ def experiment(
...
+                statistics on the test_set,
+            `training_stats` is a dictionary of training statistics for
```
Comment (w4nderlust) [inline ludwig/api.py:944]: training_stats -> training_stats
In general I prefer more explicit names in the docs, in the code we can name them differently for convenience and increasing visibility, but here clarity is the main concern

File: ludwig/api.py (hunk 14 above)
Comment (jimthompson5802) [reply ludwig/api.py:165]: np...I actually thought that but decided to be lazy...got caught by the professor. :-)
done.

File: ludwig/api.py (hunk 16 above)
Comment (jimthompson5802) [reply ludwig/api.py:942]: done

File: ludwig/api.py (hunk 17 above)
Comment (jimthompson5802) [reply ludwig/api.py:944]: done

File: ludwig/api.py (hunk 15 above)
Comment (jimthompson5802) [reply ludwig/api.py]: done

**Comment by NovikovEvgeny** [comment]
Hi @w4nderlust , @jimthompson5802 

Shouldn't docs pages also be changed/redeployed? - https://ludwig-ai.github.io/ludwig-docs/examples/ 

Many pages (including "examples") still use `--data_csv` flag to specify input dataset, what can be confusing...

Thanks

**Comment by w4nderlust** [comment]
That's because they are aligned with the most recent release: 0.2.2.8 which is the one you get from pypi. When we publish v0.3 on pypi the website will be updated

**Comment by NovikovEvgeny** [comment]
ok, got it! Thank you for the explanation, I wasn't sure how published docs are synchronized with versions

btw any information about plans when v0.3 is going to be published?

**Comment by w4nderlust** [comment]
It's coming very soon, likely next week ;)
